### ISSUE TITLE
syncoid removed all files on target on second run

### ISSUE THREAD

**Issue by danboid** [jimsalterjrs/sanoid#732]
I have set up a new Ubuntu 20.04 server, all the packages are up to date and its using the latest default kernel 5.4.0-105-generic and the default sanoid version, 2.0.3-2, which I realise is a version behind but the sanoid changelog doesn't indicate this issue might've been fixed, if it is a syncoid/sanoid bug instead of a ZFS bug. The source server is also running Ubuntu 20.04 with latest default kernel and sanoid version ie same versions as the target machine.

On my source server, I have a dataset for /home which has hundreds of child datasets, one for each home directory. I was able to successfully sync /home and all of its child datasets from the source to the target server using this command on the target machine:

```
# syncoid --no-sync-snap --no-privilege-elevation --recursive zfsbackup@192.168.10.2:astarray/home letome/home
```

The home dataset on the target (the letome pool) did not exist before I ran this command because that seems to be the way syncoid prefers to operate.

It took about 9 hours to transfer all the home dirs and I could see the files in the various directories were all present after the initial sync but I noticed that syncoid didn't copy the mount point which makes sense to me so after the initial sync had completed, I set the mountpoint of letome/home so that it would be mounted under /home, just as astarray/home is mounted. I didn't think this would cause a problem for subsequent runs of syncoid but thats about the only thing I changed on the source and target pools before the same syncoid command got run again last night via a cron job and this morning I discover that although all of the home dirs on the target are still present (as are the snapshots) all of the home dirs are empty. There were no files or sub dirs left on the target pool after re-running the above syncoid command via cron, only the base home dirs.

If this is a ZFS bug I do have the option of re-installing using a Ubuntu server 22.04 beta or daily iso.

**Comment by danboid** [comment]
I explained my problem in #openzfs and someone suggested I run `zfs mount -a` on the target machine which made all of my files re-appear which begs two question.

What could've caused every dataset on my target machine to get unmounted?

and

syncoid seems to work even when every dataset on the target exists but is umounted. That's great if it can work like that but I thought it would've errored out?

**Comment by jimsalterjrs** [comment]
> What could've caused every dataset on my target machine to get unmounted whilst the pool was still imported?

Sometimes mounting and unmounting during replication gets weird. In the initial versions of syncoid, it actually manually unmounted every filesystem on the target, just to make sure that people didn't accidentally put data into a woodchipper. But I found that getting them remounted afterward wasn't very consistent, so I stopped doing that. Now, syncoid leaves the mounted/unmounted status of datasets alone (but ZFS may still do things with them itself).

> syncoid seems to work even when every dataset on the target exists but is unmounted. That's great if it can work like that but I thought it would've errored out?

ZFS replication does not in any way require either source or target dataset or zvol to be mounted. The replication happens at a lower level than Unix filesystem mounts; as long as the pools are imported on both ends you're good to go.
